extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -201,41 +267,46 @@ private boolean ifEmptyOverPresentInWindowOperstors(
         operatorFactoryList = new ArrayList<>();
       } else if (of instanceof NaivePartitioningOperatorFactory) {
         if (((NaivePartitioningOperatorFactory) of).getPartitionColumns().isEmpty()) {
+          // TODO: This logic need to be revamped in the future. We probably don't need to handle empty over() cases separately.
           operatorList.clear();
           operatorList.add(originalQuery.getOperators());
-          return true;
+          return operatorList;
         }
       }
     }
-    return false;
+    return operatorList;
   }
 
   private ShuffleSpec findShuffleSpecForNextWindow(List<OperatorFactory> operatorFactories, int maxWorkerCount)
   {
     NaivePartitioningOperatorFactory partition = null;
     NaiveSortOperatorFactory sort = null;
-    List<KeyColumn> keyColsOfWindow = new ArrayList<>();
     for (OperatorFactory of : operatorFactories) {
       if (of instanceof NaivePartitioningOperatorFactory) {
         partition = (NaivePartitioningOperatorFactory) of;
       } else if (of instanceof NaiveSortOperatorFactory) {
         sort = (NaiveSortOperatorFactory) of;
       }
     }
-    Map<String, ColumnWithDirection.Direction> colMap = new HashMap<>();
+
+    Map<String, ColumnWithDirection.Direction> sortColumnsMap = new HashMap<>();
     if (sort != null) {
       for (ColumnWithDirection sortColumn : sort.getSortColumns()) {
-        colMap.put(sortColumn.getColumn(), sortColumn.getDirection());
+        sortColumnsMap.put(sortColumn.getColumn(), sortColumn.getDirection());
       }
     }
-    assert partition != null;
-    if (partition.getPartitionColumns().isEmpty()) {
+
+    if (partition == null || partition.getPartitionColumns().isEmpty()) {
+      // If operatorFactories doesn't have any partitioning factory, then we should keep the shuffle spec from previous stage.
+      // This indicates that we already have the data partitioned correctly, and hence we don't need to do any shuffling.
       return null;
     }
+
+    List<KeyColumn> keyColsOfWindow = new ArrayList<>();
```

Comment:
it seems like there should be a method which translates `sort.getSortColumns()` to `keyColsOfWindow`

Reply:
I can work on this as part of creating a separate class as discussed earlier:
```
class WndStage {
  PartitionOperator partitionOperator;
  SortOperator sortOperator;
  List<Operator>  workOperators;
}
```

This logic can be a method inside this class I think. Will get to this in a future PR soon.

Reply:
probably; or even outside it....we have at least 5 classes to describe sorting it would be great to use less :D so that these conversions are not necessary

but sure - it can be addressed later